### PR TITLE
Stop changes to a dupped `ActionController::Parameters` mutating the original

### PR DESCRIPTION
**Pull request by timrogers**
When `ActionController::Parameters` is duplicated with `#dup`, it doesn't create a duplicate of the instance variables (e.g. `@parameters`) but rather maintains the reference (see http://ruby-doc.org/core-2.3.1/Object.html). Given that the parameters object is often manipulated as if it were a hash (e.g. with `#delete` and similar methods), this leads to unexpected behaviour, like the following:

```
params = ActionController::Parameters.new(foo: "bar")
duplicated_params = params.dup
duplicated_params.delete(:foo)

params == duplicated_params
```

This fixes the bug by defining a private `#initialize_copy` method, used internally by `#dup`, which makes a copy of `@parameters`.

### CODE REVIEW

**Comment by bluehallu** [comment]
Isn't 

`new_params = params.except :foo`

The better approach?

**Comment by timrogers** [comment]
@bluehallu Using `#except` is a partial solution, but to me the `#dup` behaviour seems pretty broken! Keen to see what @chancancode thinks though.

**Comment by timrogers** [comment]
I must say, I'm confused by the intention of the existing `#dup` method - wouldn't `@permitted` get copied anyway?

File: actionpack/lib/action_controller/metal/strong_parameters.rb [hunk 1]
```diff
@@ -783,6 +783,11 @@ def hash_filter(params, filter)
           end
         end
       end
+
+      def initialize_copy(source)
+        super
+        @parameters = source.instance_variable_get(:@parameters).dup
```
Comment (matthewd) [inline actionpack/lib/action_controller/metal/strong_parameters.rb]: At this point, we can just use `@parameters = @parameters.dup`, I think?

**Comment by matthewd** [comment]
> I must say, I'm confused by the intention of the existing #dup method - wouldn't @permitted get copied anyway?

I imagine that may not have been true while this class inherited from Hash.

If you can confirm it works the same without it (and that we have test coverage to back that up), this seems like a good opportunity to clean up, before we make things look even more complicated, with a custom `dup` *and* an `initialize_copy`.

**Comment by timrogers** [comment]
@matthewd Agreed. Can you take another look? I've removed `#dup` altogether and replaced it with a simpler `#initialize_copy`. Some tests I've added confirm that `@permitted` is preserved.

**Comment by matthewd** [comment]
Thanks! :heart:

**Comment by rafaelfranca** [comment]
Backported in 702db0c

**Comment by rafaelfranca** [comment]
Backported in 702db0c

**Pull request merged**